src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+                let (digest_algorithm, group_id) = match sec_alg {
+                    SecAlg::ECDSAP256SHA256 => {
+                        (MessageDigest::sha256(), Nid::X9_62_PRIME256V1)
+                    }
+                    SecAlg::ECDSAP384SHA384 => {
+                        (MessageDigest::sha384(), Nid::SECP384R1)
+                    }
+                    _ => unreachable!(),
+                };
+
+                let group = EcGroup::from_curve_name(group_id)
+                    .expect("should not fail");
+                let mut ctx = BigNumContext::new().expect("should not fail");
+
+                // Add 0x4 identifier to the ECDSA pubkey as expected by openssl.
+                let public_key = dnskey.public_key().as_ref();
+                let mut key = Vec::with_capacity(public_key.len() + 1);
+                key.push(0x4);
+                key.extend_from_slice(public_key);
+                let point = EcPoint::from_bytes(&group, &key, &mut ctx)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let public_key = EcKey::from_public_key(&group, &point)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                public_key
+                    .check_key()
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+
+                Ok(PublicKey::EcDsa(
+                    digest_algorithm,
+                    public_key,
+                    dnskey.flags(),
+                ))
+            }
+            SecAlg::ED25519 => {
+                let public_key = PKey::public_key_from_raw_bytes(
+                    dnskey.public_key().as_ref(),
+                    Id::ED25519,
+                )
+                .map_err(|_| AlgorithmError::InvalidData)?;
+                Ok(PublicKey::NoDigest(public_key, dnskey.flags()))
+            }
+            _ => Err(AlgorithmError::Unsupported),
+        }
+    }
+
+    /// Verify a signature.
+    pub fn verify(
+        &self,
+        signed_data: &[u8],
+        signature: &[u8],
+    ) -> Result<(), AlgorithmError> {
+        let valid = match self {
+            PublicKey::Default(digest_algorithm, public_key, _) => {
+                let mut verifier =
+                    Verifier::new(*digest_algorithm, public_key.as_ref())
+                        .map_err(|_| AlgorithmError::InvalidData)?;
+                verifier
+                    .verify_oneshot(signature, signed_data)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+            PublicKey::NoDigest(public_key, _) => {
+                let mut verifier =
+                    Verifier::new_without_digest(public_key.as_ref())
+                        .map_err(|_| AlgorithmError::InvalidData)?;
+                verifier
+                    .verify_oneshot(signature, signed_data)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+            PublicKey::EcDsa(digest_algorithm, public_key, _) => {
```

Comment:
The original `src/validate.rs` function `verify_signed_data()` would raise an unsupported error for the ED448 case but this code does not. Is that intentional?